pkg/middleware/stored_session.go

```diff
@@ -105,51 +110,48 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
 	return session, nil
 }
 
-// refreshSessionIfNeeded will attempt to refresh a session if the session
-// is older than the refresh period.
+// refreshSessionIfNeeded will attempt to refresh a session
 // It is assumed that if the provider refreshes the session, the session is now
 // valid.
 // If the session requires refreshing but the provider does not refresh it,
 // we must validate the session to ensure that the returned session is still
 // valid.
 func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
-	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
-		// Refresh is disabled or the session is not old enough, do nothing
+	if !s.refresh || session == nil {
+		// Refresh is disabled or there is no session, do nothing
 		return nil
 	}
 
-	logger.Printf("Refreshing %s old session cookie for %s (refresh after %s)", session.Age(), session, s.refreshPeriod)
-	refreshed, err := s.refreshSessionWithProvider(rw, req, session)
-	if err != nil {
-		return err
-	}
-
-	if !refreshed {
-		// Session wasn't refreshed, so make sure it's still valid
+	if session.ExpiresOn != nil && session.ExpiresOn.After(time.Now()) || session.RefreshToken == "" {
```

Comment:
If we have adjusted the session expiry, this will cause revalidation to happen earlier

Reply:
No, revalidation will happen every time session has not expired (ExpiresOn *after* now, i.e. not yet expired). This is called every time, as in the previous logic. Due to refresh with provider being called, control order is reversed vs previous logic order: 

https://github.com/oauth2-proxy/oauth2-proxy/pull/915/files#diff-e5dfd50d27a8bea3c9c491ec3ac5b4c6874f5d4d0c025c3863fcbb04eb2b70edL127

https://github.com/oauth2-proxy/oauth2-proxy/pull/915/files#diff-afad91fa3056c5ff14edd03531c76ccafad6af201d59211c2ae95ce1dae43681L164